src/spectral.ts

```diff
@@ -25,9 +26,9 @@ export class Spectral {
     this.resolver = opts && opts.resolver ? opts.resolver : new Resolver();
   }
 
-  public async run(target: object): Promise<IRuleResult[]> {
+  public async run(target: object, parserMeta?: IParserMeta): Promise<IRuleResult[]> {
```

Comment:
`parserMeta` is needed to calculate lines and columns for given JSON Path.

Ideally, we could have a single argument accepting `IParserResult`, but that would be breaking and would force users to use `parseWithPointers` functions.

If `parserMeta` is provided, range property (lines and columns) is attached to a single result and hence can be displayed in the output (CLI) or consumed programmatically (API).

@philsturgeon @tbarn your thoughts on that?
We will need to document it for sure.

Reply:
What is this thing for?

Reply:
Also, `IRange` (range property) follows [LSP](https://microsoft.github.io/language-server-protocol/specification#range) standard and hence lines and characters (columns) are 0-based.

Reply:
@philsturgeon It's needed by `getJsonPathForPosition`

Reply:
I know the above looks slightly weird, but it preserves the old behavior.
It simply degrades gracefully - if no `parserMeta` is provided, all validation results will still have all the properties except range (and source - name of the file).

We could have a single argument, but we'd force user to use `parseWithPointers` function in such case, which is something we most likely want to avoid.
The other possible solution is to have another method, i.e. `runParsed` or similar, but IMHO this would introduce even bigger confusion to end-users.

@tbarn @philsturgeon @marbemac thoughts?
Note, The solution is also briefly described in the readme above.
Let me know what your thoughts are.

Reply:
I'd vote for keeping run as is, with a single argument, and simply `parseWithPointers(safeStringify(target))`. Basically re-stringify it if an object is passed in. Slight trade off re a potential extra parse but not a huge deal for the linting use case imho.

Thoughts?

Reply:
Yeah, sounds good to me.
What about S^2 (ya know what I mean I hope)? I'd like to avoid reparsing as much as possible in this particular case as this may degrade performance.

Reply:
Or did you mean to have the following implementation

```
run(target: IParserAstResult | object) {
  // some check here to see if we deal with parsed content or not.
}
```

Reply:
What about something like this?

Not sure we even need to document the `IParserAstResult` right now - can use internally and from studio for now.

```
run(target: IParserAstResult | object | string) {
  let parsed = target;
  if (!isParserAstResult(target)) {
    parsed = parseWithPointers(safeStringify(target));
  }

  // ...resolve parsed, etc
}
```

Reply:
The above looks good to me.

Reply:
Yeah having it there but hidden from docs is fine with me.